core/src/main/java/hudson/model/Slave.java

```diff
@@ -340,6 +343,23 @@ public String getLabelString() {
         return Util.fixNull(label).trim();
     }
 
+    private transient Set<LabelAtom> previouslyAssignedLabels = ConcurrentHashMap.newKeySet();
+
+    /**
+     * @return the labels to be trimmed for this node. This includes current and previous labels that were applied before the last call to this method.
+     */
+    @Override
+    @NonNull
+    @Restricted(NoExternalUse.class)
+    public Set<LabelAtom> drainLabelsToTrim() {
+        var result = new HashSet<>(super.drainLabelsToTrim());
+        synchronized (previouslyAssignedLabels) {
```

Comment:
Synchronizing on a concurrent set?

Reply:
This is to ensure atomicity of getting all existing values + clearing set.

Reply:
I mean, also synchronize on it here.